src/org/traccar/database/PermissionsManager.java

```diff
@@ -298,52 +305,79 @@ public void checkRegistration(long userId) {
+        } else if (object.equals(User.class) || object.equals(ManagedUser.class)) {
+            checkUser(userId, objectId);
+        } else if (object.equals(Geofence.class)) {
+            manager = Context.getGeofenceManager();
+        } else if (object.equals(Attribute.class)) {
+            manager = Context.getAttributesManager();
+        } else if (object.equals(Driver.class)) {
+            manager = Context.getDriversManager();
+        } else if (object.equals(Calendar.class)) {
+            manager = Context.getCalendarManager();
+        } else {
+            throw new IllegalArgumentException("Unknown object type");
         }
-    }
 
-    public void checkAttribute(long userId, long attributeId) throws SecurityException {
-        if (!Context.getAttributesManager().checkAttribute(userId, attributeId) && !isAdmin(userId)) {
-            checkManager(userId);
-            for (long managedUserId : getUserPermissions(userId)) {
-                if (Context.getAttributesManager().checkAttribute(managedUserId, attributeId)) {
-                    return;
+        if (manager != null) {
+            if (manager.checkItemPermission(userId, objectId) && !isAdmin(userId)) {
+                checkManager(userId);
+                for (long managedUserId : getUserPermissions(userId)) {
+                    if (manager.checkItemPermission(managedUserId, objectId)) {
+                        return;
+                    }
                 }
+                throw new SecurityException("Type " + object + " access denied");
             }
-            throw new SecurityException("Attribute access denied");
         }
     }
 
-    public void checkDriver(long userId, long driverId) throws SecurityException {
-        if (!Context.getDriversManager().checkDriver(userId, driverId) && !isAdmin(userId)) {
-            checkManager(userId);
-            for (long managedUserId : getUserPermissions(userId)) {
-                if (Context.getDriversManager().checkDriver(managedUserId, driverId)) {
-                    return;
-                }
-            }
-            throw new SecurityException("Driver access denied");
+    public void refreshAllExtendedPermissions() {
+        if (Context.getGeofenceManager() != null) {
+            Context.getGeofenceManager().refreshExtendedPermissions();
         }
+        Context.getDriversManager().refreshExtendedPermissions();
+        Context.getAttributesManager().refreshExtendedPermissions();
     }
 
-
-    public void checkCalendar(long userId, long calendarId) throws SecurityException {
-        if (!Context.getCalendarManager().checkCalendar(userId, calendarId) && !isAdmin(userId)) {
-            checkManager(userId);
-            for (long managedUserId : getUserPermissions(userId)) {
-                if (Context.getCalendarManager().checkCalendar(managedUserId, calendarId)) {
-                    return;
+    public void refreshPermissions(Map<String, Long> entity) {
```

Comment:
I think I had this comment already, but I think it would be better to take `Permission` as an argument here. So, we can compare actual classes which should be faster than constructing many string and comparing them.